### PR TITLE
Add support for OpenAPI schema transformers

### PR DESCRIPTION
**Pull request by captainsafia**
This PR adds support for the OpenAPI schema transformers API to the built-in generator.

Schema Transformer:

* [`src/OpenApi/src/Services/OpenApiOptions.cs`](diffhunk://#diff-1e67d8ed05b45af61307238fa24b706d89fc8c6ddba73bac108af00c26b46dbeR17): Added a new `UseSchemaTransformer` method that allows users to register a delegate as a schema transformer. This method adds the provided transformer delegate to the `SchemaTransformers` list. [[1]](diffhunk://#diff-1e67d8ed05b45af61307238fa24b706d89fc8c6ddba73bac108af00c26b46dbeR17) [[2]](diffhunk://#diff-1e67d8ed05b45af61307238fa24b706d89fc8c6ddba73bac108af00c26b46dbeR93-R105)
* [`src/OpenApi/perf/Microbenchmarks/TransformersBenchmark.cs`](diffhunk://#diff-9fc86114a6c135b21594047b7df9782e874098915df63e3f61ef2f483a91b79cR68-R89): A new `SchemaTransformer_Setup` method has been added. This method sets up a new schema transformer that adds an extension to the schema based on the context. A new benchmark method, `SchemaTransformer`, has also been added. [[1]](diffhunk://#diff-9fc86114a6c135b21594047b7df9782e874098915df63e3f61ef2f483a91b79cR68-R89) [[2]](diffhunk://#diff-9fc86114a6c135b21594047b7df9782e874098915df63e3f61ef2f483a91b79cR108-R113)

|            Method | TransformerCount |     Mean |    Error |   StdDev |     Op/s |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------ |----------------- |---------:|---------:|---------:|---------:|-------:|------:|------:|----------:|
| SchemaTransformer |               10 | 10.01 us | 0.096 us | 0.085 us | 99,938.4 | 0.1221 |     - |     - |     20 KB |
| SchemaTransformer |              100 | 13.01 us | 0.141 us | 0.110 us | 76,890.6 | 0.1831 |     - |     - |     25 KB |
| SchemaTransformer |             1000 | 40.00 us | 0.289 us | 0.242 us | 25,002.0 | 0.4883 |     - |     - |     82 KB |

Additional Changes:

* [`src/OpenApi/src/PublicAPI.Unshipped.txt`](diffhunk://#diff-8d0a0675439c09083194d712f368bc9905ed7e2a7595b37f668f2a4d2c79416dR14-R27): Added new entries related to the `OpenApiSchemaTransformerContext` class and the `UseSchemaTransformer` method in the `OpenApiOptions` class.
* [`src/OpenApi/perf/Microbenchmarks/TransformersBenchmark.cs`](diffhunk://#diff-9fc86114a6c135b21594047b7df9782e874098915df63e3f61ef2f483a91b79cR7): Imported the `Microsoft.OpenApi.Any` namespace.

### CODE REVIEW

File: src/OpenApi/src/Services/OpenApiDocumentService.cs [hunk 1]
```diff
@@ -99,7 +99,7 @@ internal OpenApiInfo GetOpenApiInfo()
     /// the object to support filtering each
     /// description instance into its appropriate document.
     /// </remarks>
-    private OpenApiPaths GetOpenApiPaths(HashSet<OpenApiTag> capturedTags)
+    private async Task<OpenApiPaths> GetOpenApiPathsAsync(HashSet<OpenApiTag> capturedTags, CancellationToken cancellationToken = default)
```
Comment (martincostello) [inline src/OpenApi/src/Services/OpenApiDocumentService.cs:102]: Is making all these private methods have the cancellation token be optional just setting the code up for a future edit to accidentally forget to pass it along? Should they just all be required except for on any public methods?

File: src/OpenApi/src/Services/OpenApiOptions.cs [hunk 2]
```diff
@@ -89,4 +90,17 @@ public OpenApiOptions UseOperationTransformer(Func<OpenApiOperation, OpenApiOper
         DocumentTransformers.Add(new DelegateOpenApiDocumentTransformer(transformer));
         return this;
     }
+
+    /// <summary>
+    /// Registers a given delegate as an schema transformer on the current <see cref="OpenApiOptions"/> instance.
```
Comment (martincostello) [inline src/OpenApi/src/Services/OpenApiOptions.cs:95]:
```suggestion
    /// Registers a given delegate as a schema transformer on the current <see cref="OpenApiOptions"/> instance.
```

File: src/OpenApi/src/Services/OpenApiOptions.cs [hunk 3]
```diff
@@ -89,4 +90,17 @@ public OpenApiOptions UseOperationTransformer(Func<OpenApiOperation, OpenApiOper
...
+    /// </summary>
+    /// <param name="transformer">The delegate representing the schema transformer.</param>
+    /// <returns>The <see cref="OpenApiOptions"/> instance for further customization.</returns>
+    public OpenApiOptions UseSchemaTransformer(Func<OpenApiSchema, OpenApiSchemaTransformerContext, CancellationToken, Task> transformer)
+    {
+        ArgumentNullException.ThrowIfNull(transformer, nameof(transformer));
```
Comment (martincostello) [inline src/OpenApi/src/Services/OpenApiOptions.cs:101]:
```suggestion
        ArgumentNullException.ThrowIfNull(transformer);
```

File: src/OpenApi/src/Transformers/OpenApiSchemaTransformerContext.cs [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+{
+    /// <summary>
+    /// Gets the name of the associated OpenAPI document.
+    /// </summary>
+    public required string DocumentName { get; init; }
+
+    /// <summary>
+    /// Gets the <see cref="Type" /> associated with the current <see cref="OpenApiSchema"/>
+    /// </summary>
+    /// <value></value>
```
Comment (martincostello) [inline src/OpenApi/src/Transformers/OpenApiSchemaTransformerContext.cs:22]:
```suggestion
```

File: src/OpenApi/src/Transformers/OpenApiSchemaTransformerContext.cs [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    public required Type Type { get; init; }
+
+    /// <summary>
+    /// Gets the <see cref="ApiParameterDescription"/>  associated with target schema.
```
Comment (martincostello) [inline src/OpenApi/src/Transformers/OpenApiSchemaTransformerContext.cs:26]:
```suggestion
    /// Gets the <see cref="ApiParameterDescription"/> associated with the target schema.
```

**Comment by davidfowl** [comment]
Can these transformers access the IServiceProvider?

File: src/OpenApi/src/Services/OpenApiOptions.cs [hunk 6]
```diff
@@ -14,6 +14,7 @@ namespace Microsoft.AspNetCore.OpenApi;
 public sealed class OpenApiOptions
 {
     internal readonly List<IOpenApiDocumentTransformer> DocumentTransformers = [];
+    internal readonly List<Func<OpenApiSchema, OpenApiSchemaTransformerContext, CancellationToken, Task>> SchemaTransformers = [];
```
Comment (JamesNK) [inline src/OpenApi/src/Services/OpenApiOptions.cs:17]: What are scenarios where someone could have an async schema transformer?

Comment (JamesNK) [reply src/OpenApi/src/Services/OpenApiOptions.cs:17]: Is XML comments file from disk using async and then applying changes to schema an example?

**Comment by JamesNK** [comment]
> Can these transformers access the IServiceProvider?

There is `OpenApiSchemaTransformerContext.ApplicationServices`.

**Comment by captainsafia** [comment]
> Can these transformers access the IServiceProvider?

In addition to the property on the context object @JamesNK mentioned above, we also have a generic API for supporting resolving activated instances of a transformer `OpenApiOptions.UseTransformer<T>`. To keep the API small, we've only exposed it for document transformers at the moment but it's a pattern we can expand to other transformers as the need arises (e.g. there's an API proposal to add it. for operation transformers).

File: src/OpenApi/src/Services/OpenApiOptions.cs (hunk 6 above)
Comment (captainsafia) [reply src/OpenApi/src/Services/OpenApiOptions.cs:17]: XML comments support will be handled separately in this implementation, although theoretically there's nothing stopping someone from rolling out an implementation on their own or reading additional metadata from some non-XML file.

File: src/OpenApi/src/Transformers/OpenApiSchemaTransformerContext.cs [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+/// </summary>
+public sealed class OpenApiSchemaTransformerContext
+{
+    /// <summary>
+    /// Gets the name of the associated OpenAPI document.
+    /// </summary>
+    public required string DocumentName { get; init; }
+
+    /// <summary>
+    /// Gets the <see cref="Type" /> associated with the current <see cref="OpenApiSchema"/>
```
Comment (BrennanConroy) [inline src/OpenApi/src/Transformers/OpenApiSchemaTransformerContext.cs:20]:
```suggestion
    /// Gets the <see cref="Type" /> associated with the current <see cref="OpenApiSchema"/>.
```

**Review by BrennanConroy — APPROVED** [review]
Nit: Should add a test for checking DocumentName and ApplicationServices.

File: src/OpenApi/test/Transformers/SchemaTransformerTests.cs [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+    [Fact]
+    public async Task SchemaTransformer_WithDescriptionOnlyModifiesParameter()
+    {
+        var builder = CreateBuilder();
+
+        builder.MapPost("/todo", (Todo todo) => { });
+        builder.MapGet("/todo", () => new Todo(1, "Item1", false, DateTime.Now));
+
+        var options = new OpenApiOptions();
+        options.UseSchemaTransformer((schema, context, cancellationToken) =>
```
Comment (halter73) [inline src/OpenApi/test/Transformers/SchemaTransformerTests.cs:114]: Can we add a test the verifies we flow the `CancellationToken` all the way through?

**Review by halter73 — APPROVED** [review]
Did we discuss `Use...Transformer` vs `Add...Transformer` in API review? I only ask because we went with `AddEndpointFilter` over `UseEndpointFilter` in .NET 7, and this feels even more addy so-to-speak since there can be multiples of each type of transformer and there's no `next` callback unlike filters and middleware.

**Comment by captainsafia** [comment]
> Did we discuss `Use...Transformer` vs `Add...Transformer` in API review? I only ask because we went with `AddEndpointFilter` over `UseEndpointFilter` in .NET 7, and this feels even more addy so-to-speak since there can be multiples of each type of transformer and there's no `next` callback unlike filters and middleware.

We discussed the similarities/differences between transformers and middleware/filters. Looking back at [the notes](https://github.com/dotnet/aspnetcore/issues/54650), it doesn't look like we discussed `Use` vs `Add` specifically.

I have a slight preference towards `Add` myself. We can discuss this rename at the upcoming review around adding an interface type for operation transformers.

**Pull request merged**